src/charm.py

```diff
@@ -19,6 +24,169 @@ class KratosCharm(CharmBase):
+        self.framework.observe(self.database.on.database_created, self._on_database_changed)
+        self.framework.observe(self.database.on.endpoints_changed, self._on_database_changed)
+
+    @property
+    def _pebble_layer(self) -> Layer:
+        pebble_layer = {
+            "summary": "kratos layer",
+            "description": "pebble config layer for kratos",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Kratos Operator layer",
+                    "startup": "enabled",
+                    "command": f"kratos serve all --config {self._config_file_path}",
+                }
+            },
+            "checks": {
+                "ready": {
+                    "override": "replace",
+                    "http": {"url": "http://localhost:4434/admin/health/ready"},
+                },
+                "alive": {
+                    "override": "replace",
+                    "http": {"url": "http://localhost:4434/admin/health/alive"},
+                },
+            },
+        }
+        return Layer(pebble_layer)
+
+    @property
+    def _config(self) -> str:
+        db_info = self._get_database_relation_info()
+        config = {
+            "log": {"level": "trace"},
+            "identity": {
+                "default_schema_id": "default",
+                "schemas": [{"id": "default", "url": f"file://{self._identity_schema_file_path}"}],
+            },
+            "selfservice": {
+                "default_browser_return_url": "http://127.0.0.1:9999/",
+                "flows": {
+                    "registration": {
+                        "enabled": True,
+                        "ui_url": "http://127.0.0.1:9999/registration",
+                    }
+                },
+            },
+            "dsn": f"postgres://{db_info['username']}:{db_info['password']}@{db_info['endpoints']}/postgres",
+            "courier": {
+                "smtp": {
+                    # TODO: dynamic connection uri through charm config
+                    "connection_uri": "smtps://test:test@mailslurper:1025/?skip_ssl_verify=true"
+                }
+            },
+        }
+
+        return yaml.dump(config)
+
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer and kratos config if changed."""
+        if not self._container.get_plan().to_dict():
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.push(self._config_file_path, self._config, make_dirs=True)
+            with open("src/identity.default.schema.json", encoding="utf-8") as schema_file:
+                schema = schema_file.read()
+                self._container.push(self._identity_schema_file_path, schema, make_dirs=True)
+            self._container.add_layer(self._container_name, self._pebble_layer, combine=True)
```

Comment:
You are adding the layer only once. It makes sense as long as there are no configurable options, but once we add any, the layer will need to be added and container replanned not only when the plan is empty.

Reply:
Right now the layer doesn't change at all. We could update this in the future if the layer changes